cvmfs/session_context.cc

```diff
@@ -409,30 +404,27 @@ bool SessionContext::DoUpload(const SessionContext::UploadJob* job) {
 
 void* SessionContext::UploadLoop(void* data) {
   SessionContext* ctx = reinterpret_cast<SessionContext*>(data);
+  UploadJob *job;
 
-  int64_t jobs_processed = 0;
-  while (!ctx->ShouldTerminate()) {
-    while (jobs_processed < ctx->NumJobsSubmitted()) {
-      UploadJob* job = ctx->upload_jobs_->Dequeue();
+  do {
+    job = ctx->upload_jobs_->Dequeue();
+    if (job != &terminator_) {
       if (!ctx->DoUpload(job)) {
         PANIC(kLogStderr,
               "SessionContext: could not submit payload. Aborting.");
       }
       job->result->Set(true);
       delete job->pack;
       delete job;
-      jobs_processed++;
     }
-    if (ctx->queue_was_flushed_.IsEmpty()) {
+    if (ctx->upload_jobs_->IsEmpty() && ctx->queue_was_flushed_.IsEmpty()) {
```

Comment:
The logic `if (IsEmpty()) Enqueue();` smells like a race condition. I realize that it was there before, I need another moment to think it through.

Reply:
Agreed that the code needs more fixing.  I don't think this PR adds any race conditions that weren't already present, but it does at least remove a CPU spinloop.  Could fixing the race condition(s) be deferred for a later fix?

Reply:
Boy, these classes need a rewrite.

If I'm not mistaken, we can get rid of `queue_was_flushed_` entirely.  It seems to me not to serve any useful purpose.  The way it is implemented, it blocks the `WaitForUpload()` call until a submitted object pack in flight is sent and submitted to the gateway.  However, that logic only starts as of the second `WaitForUpload()` call because the `queue_was_flushed_` is initially `true`.  Also, the upload queue works on entire object packs -- so as long as an object pack is not full (and hence not dispatched), `WaitForUpload()` doesn't wait.

The barrier for the gateway uploader is actually not `WaitForUpload()` but `FinalizeSession()`.  Finalizing the session will end up in `SessionContextBase::Finalize`, which looks at the currently open object pack and closes out the tab itself.

Let's try (in this PR) removing `queue_was_flushed_` and just returning from `WaitForUpload()`.

Reply:
Added an extra commit to remove `queue_was_flushed_` completely, as described.  I put you as the commit author since my contribution here is merely secretarial.